File: src/models/axial_steady.py
# src/models/axial_steady.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path
from src.props.props_water import water_props_T  # T-dependent (educational fits)

# ---------------- Base inputs ----------------
L = 6.0        # m  (channel length)
D_h = 0.012    # m  (hydraulic diameter of equivalent subchannel)
T_in = 560.0   # K  (inlet coolant temperature)
P = 10e6       # Pa (channel pressure)
m_dot = 2.0    # kg/s (mass flow rate)

# Choose heat rate (use 60e3 for "cooler" demo; 350e3 for "hot" base)
q_lin = 25e3  # W/m

# Pipe roughness for Churchill friction (m): ~5e-6 smooth metal
eps = 0

# Discretization
N = 200
x = np.linspace(0, L, N + 1)
dx = x[1] - x[0]

# Geometry
P_wet = np.pi * D_h
A_flow = np.pi * (D_h / 2) ** 2
q_pp = q_lin / P_wet  # W/m^2

# ---------------- Arrays ----------------
T_c = np.zeros_like(x); T_c[0] = T_in
T_w = np.zeros_like(x)
h   = np.zeros_like(x)
Re  = np.zeros_like(x)
Nu  = np.zeros_like(x)
f   = np.zeros_like(x)
dP  = np.zeros_like(x)  # cumulative Pa

# ---------------- Helpers ----------------
def h_from_props(rho, mu, k, cp, D_h, m_dot, A_flow):
    """Dittus–Boelter heat transfer using local props."""
    v  = m_dot / (rho * A_flow)
    Re = rho * v * D_h / mu
    Pr = cp * mu / k
    Nu = 0.023 * (Re**0.8) * (Pr**0.4)
    h  = Nu * k / D_h
    return h, Re, Nu, v

def f_churchill(Re, D_h, eps):
    """
    Churchill (1977) friction factor, valid for all Re and roughness.
    eps: absolute roughness [m]
    """
    if Re < 1e-12:
        return 0.0
    A = (2.457 * np.log((7.0/Re)**0.9 + 0.27*(eps/D_h)))**16
    B = (37530.0 / Re)**16
    return 8.0 * ((8.0/Re)**12 + 1.0/((A + B)**1.5))**(1.0/12.0)

def tsat_water(P_pa: float) -> float:
    """
    Saturation temperature for water at pressure P (Pa).
    Tries CoolProp; else returns an educational placeholder near 10 MPa.
    """
    try:
        import CoolProp.CoolProp as CP
        return CP.PropsSI("T", "P", P_pa, "Q", 0, "Water")
    except Exception:
        return 584.0  # ~Tsat at 10 MPa (K), for visualization only

# ---------------- Axial march with film-T iteration ----------------
# Inlet film properties & initial wall estimate
props0 = water_props_T(T_in, P)
h_guess, Re0, Nu0, v0 = h_from_props(
    rho=props0["rho"], mu=props0["mu"], k=props0["k"], cp=props0["cp"],
    D_h=D_h, m_dot=m_dot, A_flow=A_flow
)
T_w[0] = T_c[0] + q_pp / h_guess
h[0], Re[0], Nu[0] = h_guess, Re0, Nu0
f[0] = f_churchill(Re[0], D_h, eps)

for i in range(N):
    # 1) Bulk energy balance (cp at current film estimate)
    props_i = water_props_T(0.5*(T_c[i] + T_w[i]), P)
    T_c[i+1] = T_c[i] + (q_lin / (m_dot * props_i["cp"])) * dx

    # 2) Film-temperature fixed-point iteration at node i+1
    T_c_loc = T_c[i+1]
    T_w_loc = T_c_loc + q_pp / max(h[i], 1e-3)  # start from previous h
    pr = props_i
    for _ in range(3):  # few passes are sufficient here
        T_f = 0.5 * (T_c_loc + T_w_loc)
        pr  = water_props_T(T_f, P)
        h_loc, Re_loc, Nu_loc, v_loc = h_from_props(
            rho=pr["rho"], mu=pr["mu"], k=pr["k"], cp=pr["cp"],
            D_h=D_h, m_dot=m_dot, A_flow=A_flow
        )
        T_w_loc = T_c_loc + q_pp / h_loc

    # Store locals
    T_w[i+1] = T_w_loc
    h[i+1]   = h_loc
    Re[i+1]  = Re_loc
    Nu[i+1]  = Nu_loc
    f[i+1]   = f_churchill(Re_loc, D_h, eps)

    # 3) Local pressure gradient with updated rho, v
    dPdx = f[i+1] * (pr["rho"] * v_loc**2 / 2.0) / D_h
    dP[i+1] = dP[i] + dPdx * dx

# ---------------- Margin to boiling ----------------
T_sat = tsat_water(P)                 # K (assume constant along channel)
DeltaT_sat = T_sat - T_w              # K
min_margin = float(DeltaT_sat.min())
ix_min = int(np.argmin(DeltaT_sat))
x_min = float(x[ix_min])

# ---------------- Output paths ----------------
base = Path(__file__).resolve().parents[2]
data_dir = base / "data"; data_dir.mkdir(exist_ok=True, parents=True)
fig_dir  = base / "figures"; fig_dir.mkdir(exist_ok=True, parents=True)

# ---------------- Save CSV ----------------
out = pd.DataFrame({
    "x_m": x,
    "T_c_K": T_c,
    "T_w_K": T_w,
    "h_W_m2K": h,
    "Re": Re,
    "Nu": Nu,
    "f": f,
    "dP_cum_Pa": dP,
    "T_sat_K": np.full_like(x, T_sat, dtype=float),
    "DeltaT_sat_K": DeltaT_sat
})
out.to_csv(data_dir / "axial_steady_base.csv", index=False)

# ---------------- Plots ----------------
# 1) Temperatures + Tsat
plt.figure()
plt.plot(x, T_c, label="Coolant T_c", linewidth=2)
plt.plot(x, T_w, label="Wall T_w", linewidth=2, linestyle="--")
plt.plot(x, np.full_like(x, T_sat), label="T_sat (≈)", linewidth=1.5, linestyle="-.")
plt.xlabel("Axial position x (m)")
plt.ylabel("Temperature (K)")
plt.title("Coolant, Wall, and Saturation Temperature vs x")
plt.grid(True, linestyle=":")
plt.legend()
plt.savefig(fig_dir / "Tx_coolant_wall_Tsat.png", dpi=200, bbox_inches="tight")

# 2) Cumulative pressure drop
plt.figure()
plt.plot(x, dP/1e6, linewidth=2)
plt.xlabel("Axial position x (m)")
plt.ylabel("ΔP (MPa)")
plt.title("Cumulative Pressure Drop vs x (Churchill friction)")
plt.grid(True, linestyle=":")
plt.savefig(fig_dir / "dP_x_filmT.png", dpi=200, bbox_inches="tight")

# 3) Margin to boiling with min annotation
plt.figure()
plt.plot(x, DeltaT_sat, linewidth=2, label="ΔT_sat = T_sat - T_w")
plt.axhline(0, linestyle="--", linewidth=1)
plt.scatter([x_min], [min_margin], zorder=5)
plt.annotate(f"min {min_margin:.1f} K at x={x_min:.2f} m",
             (x_min, min_margin),
             textcoords="offset points", xytext=(8,8))
plt.xlabel("Axial position x (m)")
plt.ylabel("ΔT_sat (K)")
plt.title("Margin to Boiling vs x")
plt.grid(True, linestyle=":")
plt.legend()
plt.savefig(fig_dir / "margin_to_boiling.png", dpi=200, bbox_inches="tight")

# ---------------- Console summary ----------------
print("=== Axial Steady with Film-T + Churchill + Boiling Margin ===")
print(f"T_out = {T_c[-1]:.2f} K, T_w,out = {T_w[-1]:.2f} K, T_sat ≈ {T_sat:.2f} K")
print(f"h_in = {h[0]:.0f} W/m^2-K, h_out = {h[-1]:.0f} W/m^2-K")
print(f"Re range ~ {Re[0]:.2e} → {Re[-1]:.2e}")
print(f"Friction: Churchill, eps = {eps:g} m")
print(f"Total ΔP = {dP[-1]/1e6:.3f} MPa over L = {L} m")
print(f"Min ΔT_sat = {min_margin:.1f} K at x = {x_min:.2f} m")
print(f"CSV:    {data_dir / 'axial_steady_base.csv'}")
print(f"Figures:{fig_dir / 'Tx_coolant_wall_Tsat.png'}, {fig_dir / 'dP_x_filmT.png'}, {fig_dir / 'margin_to_boiling.png'}")

File: matlab/axial_steady.m
%% 01_axial_steady.m
% CANDU Fuel Channel – steady 1-D model with film-T properties,
% Dittus–Boelter h, Churchill friction, and margin to boiling.
% Save this file as: matlab/01_axial_steady.m  (R2016b+ for local functions)

clear; clc;

%% ---------------- Base inputs ----------------
L     = 6.0;         % m  channel length
Dh    = 0.012;       % m  hydraulic diameter
Tin   = 560.0;       % K  inlet coolant temperature
P     = 10e6;        % Pa pressure
mdot  = 2.0;         % kg/s mass flow rate
q_lin = 15.5e3;      % W/m linear heat rate (e.g., tuned for +10 K margin)
eps   = 0.0;         % m absolute roughness (0 ~ smooth metal)

N  = 200;                        % axial nodes
x  = linspace(0, L, N+1).';      % column vector
dx = x(2)-x(1);

Pwet  = pi*Dh;
Aflow = pi*(Dh/2)^2;
qpp   = q_lin / Pwet;            % W/m^2

%% ---------------- Arrays ----------------
Tc = zeros(N+1,1); Tc(1) = Tin;
Tw = zeros(N+1,1);
h  = zeros(N+1,1);
Re = zeros(N+1,1);
Nu = zeros(N+1,1);
ff = zeros(N+1,1);
dP = zeros(N+1,1);               % cumulative Pa

%% ---------------- Inlet film estimate ----------------
props0 = water_props_T(Tin, P);
[h0, Re0, Nu0, v0] = h_from_props(props0.rho, props0.mu, props0.k, props0.cp, Dh, mdot, Aflow);
Tw(1) = Tc(1) + qpp/h0;
h(1)  = h0;  Re(1)=Re0;  Nu(1)=Nu0;
ff(1) = f_churchill(Re(1), Dh, eps);

%% ---------------- Axial march with film-T iteration ----------------
for i = 1:N
    % 1) Advance coolant using cp at current film estimate
    props_i = water_props_T(0.5*(Tc(i)+Tw(i)), P);
    Tc(i+1) = Tc(i) + (q_lin/(mdot*props_i.cp))*dx;

    % 2) Film-T iteration (few fixed-point passes)
    Tc_loc = Tc(i+1);
    Tw_loc = Tc_loc + qpp/max(h(i),1e-3); % start from previous h
    pr = props_i;
    for k = 1:3
        Tf = 0.5*(Tc_loc + Tw_loc);
        pr = water_props_T(Tf, P);
        [h_loc, Re_loc, Nu_loc, v_loc] = h_from_props(pr.rho, pr.mu, pr.k, pr.cp, Dh, mdot, Aflow);
        Tw_loc = Tc_loc + qpp/h_loc;
    end

    Tw(i+1) = Tw_loc;  h(i+1) = h_loc;  Re(i+1) = Re_loc;  Nu(i+1) = Nu_loc;
    ff(i+1) = f_churchill(Re_loc, Dh, eps);

    % 3) Hydraulics
    dPdx = ff(i+1)*(pr.rho*v_loc^2/2)/Dh;
    dP(i+1) = dP(i) + dPdx*dx;
end

%% ---------------- Margin to boiling ----------------
Tsat = tsat_water(P);                 % K (CoolProp if present; else ~584 K)
DeltaT_sat = Tsat - Tw;
[min_margin, ix_min] = min(DeltaT_sat);
x_min = x(ix_min);

%% ---------------- Save CSV (data/axial_steady_base.csv) ----------------
base    = fileparts(fileparts(pwd));  % repo root if you're in matlab/
dataDir = fullfile(base,'data'); if ~exist(dataDir,'dir'), mkdir(dataDir); end
T = table(x, Tc, Tw, h, Re, Nu, ff, dP, repmat(Tsat,size(x)), DeltaT_sat, ...
    'VariableNames', {'x_m','T_c_K','T_w_K','h_W_m2K','Re','Nu','f','dP_cum_Pa','T_sat_K','DeltaT_sat_K'});
writetable(T, fullfile(dataDir,'axial_steady_base.csv'));

%% ---------------- Plots (figures/*.png) ----------------
figDir = fullfile(base,'figures'); if ~exist(figDir,'dir'), mkdir(figDir); end

% 1) Coolant, wall, Tsat
f1 = figure('Color','w');
plot(x,Tc,'LineWidth',2); hold on;
plot(x,Tw,'--','LineWidth',2);
plot(x, Tsat*ones(size(x)),'-.','LineWidth',1.5);
grid on; xlabel('Axial position x (m)'); ylabel('Temperature (K)');
title('Coolant, Wall, and Saturation Temperature vs x');
legend('Coolant T_c','Wall T_w','T_{sat} (≈)','Location','best');
exportgraphics(f1, fullfile(figDir,'Tx_coolant_wall_Tsat.png'), 'Resolution',200);

% 2) ΔP
f2 = figure('Color','w');
plot(x, dP/1e6, 'LineWidth',2);
grid on; xlabel('Axial position x (m)'); ylabel('\DeltaP (MPa)');
title('Cumulative Pressure Drop vs x (Churchill friction)');
exportgraphics(f2, fullfile(figDir,'dP_x_filmT.png'), 'Resolution',200);

% 3) Margin + annotation
f3 = figure('Color','w');
plot(x, DeltaT_sat, 'LineWidth',2); hold on;
yline(0,'--'); scatter(x_min, min_margin, 40, 'filled');
text(x_min, min_margin, sprintf('  min %.1f K @ x=%.2f m',min_margin,x_min), 'VerticalAlignment','bottom');
grid on; xlabel('Axial position x (m)'); ylabel('\DeltaT_{sat} (K)');
title('Margin to Boiling vs x');
exportgraphics(f3, fullfile(figDir,'margin_to_boiling.png'), 'Resolution',200);

%% ---------------- Console summary ----------------
fprintf('=== Axial Steady (MATLAB) – Film-T + Churchill + Margin ===\n');
fprintf('T_out = %.2f K, T_w,out = %.2f K, T_sat ≈ %.2f K\n', Tc(end), Tw(end), Tsat);
fprintf('h_in = %.0f W/m^2-K, h_out = %.0f W/m^2-K\n', h(1), h(end));
fprintf('Re range ~ %.2e → %.2e\n', Re(1), Re(end));
fprintf('Friction: Churchill, eps = %g m\n', eps);
fprintf('Total ΔP = %.3f MPa over L = %.1f m\n', dP(end)/1e6, L);
fprintf('Min ΔT_sat = %.1f K at x = %.2f m\n', min_margin, x_min);
fprintf('CSV:    %s\n', fullfile(dataDir,'axial_steady_base.csv'));
fprintf('Figures:%s\n        %s\n        %s\n', ...
    fullfile(figDir,'Tx_coolant_wall_Tsat.png'), ...
    fullfile(figDir,'dP_x_filmT.png'), ...
    fullfile(figDir,'margin_to_boiling.png'));

%% ---------------- Local functions ----------------
function [h, Re, Nu, v] = h_from_props(rho, mu, k, cp, Dh, mdot, Aflow)
    v  = mdot/(rho*Aflow);
    Re = rho*v*Dh/mu;
    Pr = cp*mu/k;
    Nu = 0.023*(Re^0.8)*(Pr^0.4);
    h  = Nu*k/Dh;
end

function f = f_churchill(Re, Dh, eps)
    if Re < 1e-12, f = 0.0; return; end
    A = (2.457*log((7.0/Re)^0.9 + 0.27*(eps/Dh)))^16;
    B = (37530.0/Re)^16;
    f = 8.0 * ((8.0/Re)^12 + 1.0/((A+B)^1.5))^(1/12);
end

function Tsat = tsat_water(P)
% Tries CoolProp if available; else returns ~584 K at 10 MPa (illustrative).
    Tsat = 584.0;
    try
        if exist('CoolProp.PropsSI','file')
            Tsat = CoolProp.PropsSI('T','P',P,'Q',0,'Water');
        end
    catch
        % keep placeholder
    end
end

function props = water_props_T(T, P)
% Educational property fit: water-like at high pressure.
% Returns struct with rho [kg/m3], mu [Pa.s], k [W/m-K], cp [J/kg-K]
% Simple temperature trends around 550–650 K range (not D2O, illustrative).
    rho = 700 - 0.2*(T-560);         % very rough trend
    rho = max(rho, 200);
    mu  = 1.2e-4 * exp(-0.012*(T-560));  % Pa.s, decreasing with T
    k   = 0.65 - 1.0e-4*(T-560);     % W/m-K
    cp  = 5000 + 0.5*(T-560);        % J/kg-K
    props = struct('rho',rho,'mu',mu,'k',k,'cp',cp);
end
